src/transaction.rs

```diff
@@ -373,10 +375,165 @@ impl Decodable for TxOut {
     }
 }
 
+pub enum TxOutError {
+    NoBlindingKeyInAddress,
+    Upstream(secp256k1_zkp::Error),
+}
+
 impl TxOut {
+    /// Creates a new confidential output that is **not** the last one in the transaction.
+    pub fn new_not_last_confidential<R, C>(
+        rng: &mut R,
+        secp: &Secp256k1<C>,
+        value: u64,
+        address: Address,
+        asset: AssetId,
+        inputs: &[(
+            AssetId,
+            u64,
+            Generator,
+            AssetBlindingFactor,
+            ValueBlindingFactor,
+        )],
+    ) -> Result<(Self, AssetBlindingFactor, ValueBlindingFactor), TxOutError>
+    where
+        R: RngCore + CryptoRng,
+        C: Signing,
+    {
+        let out_abf = AssetBlindingFactor::new(rng);
+        let out_asset = Asset::new_confidential(secp, asset, out_abf);
+
+        let out_asset_commitment = out_asset.commitment().expect("confidential asset");
+        let out_vbf = ValueBlindingFactor::random(rng);
+        let value_commitment = Value::new_confidential(secp, value, out_asset_commitment, out_vbf);
+
+        let receiver_blinding_pk = &address
+            .blinding_pubkey
+            .ok_or(TxOutError::NoBlindingKeyInAddress)?;
+        let (nonce, shared_secret) = Nonce::new_confidential(rng, secp, receiver_blinding_pk);
+
+        let message = RangeProofMessage { asset, bf: out_abf };
+        let rangeproof = RangeProof::new(
+            secp,
+            1,
+            value_commitment.commitment().expect("confidential value"),
+            value,
+            out_vbf.0,
+            &message.to_bytes(),
+            address.script_pubkey().as_bytes(),
+            shared_secret,
+            0,
+            52,
```

Comment:
Should we document the use of these constants better? IIRC, we got them out of the documentation of libwally.

cc @apoelstra

Reply:
These come from here, yes: https://github.com/ElementsProject/libwally-core/blob/3aa27c1ba1136be5010808c70ff7e8c0f58f66a9/include/wally_elements.h#L142-143

Reply:
They're recommended by the authors of libwally, but it would be better to say why they've been chosen if we are to document them.

Reply:
The first number, 0, is used to shift the exponent, because CT supports floating-point numbers with an explicit exponent but blind mantissa. Setting this to 0 basically disables the feature, which I think represents a bad privacy/efficiency tradeoff and shouldn't be used.

Setting min-bits to 52 gives you room to represent satoshi values up to 2^52 (45MM BTC, in excess of the total 21MM supply) and was chosen to match the range of a double-width floating point value.

Reply:
> The first number, 0, is used to shift the exponent, because CT supports floating-point numbers with an explicit exponent but blind mantissa. Setting this to 0 basically disables the feature, which I think represents a bad privacy/efficiency tradeoff and shouldn't be used.

What would be a better default value? And should this API expose an optional argument to control it?

Reply:
Sorry, I worded that poorly. We **should** leave it at 0, and we shouldn't expose any way to change it.